### ISSUE TITLE
Questions about benchmark on AWS-EFA

### ISSUE THREAD

**Issue by zarzen** [ofiwg/libfabric#5694]
Hi there,

I have run `fi_rdm_tagged_bw`, `fi_rma_bw`, `fi_bw` benchmarks on two `c5n.18xl` instances placed in the same placement group. The followings are bandwidth outputs I got. I would like to know whether those results are reasonable. I am wondering why `fi_rdm_tagged_bw` outperforms the other two by that large margin. 
On the other hand, even the best performance of `fi_rdm_tagged_bw` cannot achieve 100 Gbps. 
I thought it might due to the CPU bottleneck because only one endpoint is enabled for experiments. 
But if I run two endpoints (client and server) on the same instance, it can achieve 100 Gbps. The result is presented at the last. 

```
fi_rdm_tagged_bw -p efa -s fe80::55:c0ff:fe54:7051 -P 0 fe80::54:faff:feeb:cac7
bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      20k     1.2m        0.03s     42.74       1.50       0.67
256     20k     4.8m        0.03s    161.42       1.59       0.63
1k      20k     19m         0.03s    596.41       1.72       0.58
4k      20k     78m         0.04s   2305.92       1.78       0.56
64k     2k      125m        0.02s   6189.06      10.59       0.09
1m      200     200m        0.03s   6892.86     152.12       0.01
1.5m    200     300m        0.05s   6899.89     227.96       0.00
2m      200     400m        0.06s   6868.93     305.31       0.00
4m      200     800m        0.12s   6851.09     612.21       0.00
6m      200     1.1g        0.18s   6840.32     919.76       0.00

fi_rma_bw -p efa -s fe80::55:c0ff:fe54:7051 -P 0 fe80::54:faff:feeb:cac7
bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      20k     1.2m        0.03s     43.15       1.48       0.67
256     20k     4.8m        0.03s    161.54       1.58       0.63
1k      20k     19m         0.03s    620.34       1.65       0.61
4k      20k     78m         0.03s   2346.27       1.75       0.57
64k     2k      125m        0.03s   3925.60      16.69       0.06
1m      200     200m        0.06s   3755.71     279.20       0.00

fi_bw -p efa -s fe80::55:c0ff:fe54:7051 -P 0 fe80::54:faff:feeb:cac7
bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      64      4k          0.00s     29.26       2.19       0.46
256     64      16k         0.00s    174.30       1.47       0.68
1k      64      64k         0.00s    321.25       3.19       0.31
4k      64      256k        0.00s   1432.48       2.86       0.35
64k     64      4m          0.00s   2388.56      27.44       0.04
1m      64      64m         0.03s   2209.93     474.48       0.00

# client endpoint and server endpoint on the same machine
# same EFA address
/fi_rdm_tagged_bw -p efa -s fe80::54:faff:feeb:cac7 -P 0 fe80::54:faff:feeb:cac7
bytes   iters   total       time     MB/sec    usec/xfer   Mxfers/sec
64      20k     1.2m        0.04s     36.46       1.76       0.57
256     20k     4.8m        0.08s     67.95       3.77       0.27
1k      20k     19m         0.05s    415.85       2.46       0.41
4k      20k     78m         0.07s   1199.84       3.41       0.29
64k     2k      125m        0.01s   8976.92       7.30       0.14
1m      200     200m        0.02s  12514.33      83.79       0.01
1.5m    200     300m        0.03s  10578.14     148.69       0.01
2m      200     400m        0.04s  10463.79     200.42       0.00
4m      200     800m        0.08s  10387.85     403.77       0.00
6m      200     1.1g        0.13s   9379.52     670.77       0.00

```
In conclusion there are following questions I have:
* i) Are those results expected? (`rdm_tagged_bw` performs much better than `rma_bw` and `fi_bw`)

* ii) Why two endpoints on the same machine can achieve 100 Gbps, but two endpoints on different servers cannot?

* iii) Is that possible to have two endpoints on each server to saturate 100 Gbps EFA? (I didn't see benchmarks for multiple endpoints, and I am not even able to run the `fi_multi_ep` to verify it is possible to run multiple-endpoint on EFA)

* iv) why the performance degrades after message sizes larger than 1M?

**Comment by wzamazon** [comment]
Hi, @zarzen these numbers looks normal to me.

Regarding your questions:

1.  Currently, the RMA operation of EFA is implemented by simulation, therefore the performance of rma_bw is not as good as tagged_rdm_bw. Meanwhile, fi_bw does data verification, therefore the performance is not comparable to fi_rdm_tagged_bw.

2. Two EFA end points on same machine will use SHM provider to do communication, by passing EFA, therefore the performance is better.

3. Yes, its possible to using multiple EP with EFA. You can run some MPI benchmark to verify it. For example, the mbw_mr test of OSU MPI benchmark.

**Comment by zarzen** [comment]
Hi @wzamazon, Thanks for your suggestion.
Could you point me to some examples that use multiple EP with EFA directly with `libfabric`? 
Because I would like to use `libfabric`+`efa` without MPI later. 
(I think I can find it somewhere in the openMPI code, just for convenience here, in case you know some sample code to start with)
Thanks!
